PyViCare/PyViCareServiceV2.py

```diff
@@ -0,0 +1,292 @@
+
+        Returns
+        -------
+        oauth:
+            oauth sessions object
+        """
+        oauth = OAuth2Session(self.client_id, redirect_uri=redirect_uri,scope=viessmann_scope)
+        authorization_url, _ = oauth.authorization_url(authorizeURL)
+           
+        # workaround until requests-oauthlib supports PKCE flow
+        code_verifier, code_challenge = pkce.generate_pkce_pair()
+        authorization_url += '&code_challenge=' + code_challenge + '&code_challenge_method=S256'
+        logger.debug("Auth URL is: "+authorization_url)
+
+        try:
+            header = {'Content-Type': 'application/x-www-form-urlencoded'}
+            response = requests.post(authorization_url, headers=header,auth=(username,password))
+            logger.warning("Received an HTML answer from the server during auth, this is not normal:")
+            logger.debug(response.content)
+        except requests.exceptions.InvalidSchema as e:
+            #capture the error, which contains the code the authorization code and put this in to codestring
+            codestring = "{0}".format(str(e.args[0])).encode("utf-8")
+            codestring = str(codestring)
+            match = re.search("code=(.*)&",codestring)
+            codestring=match.group(1)
+            logger.debug("Codestring : "+codestring)
+
+            # workaround until requests-oauthlib supports PKCE flow
+            resp = requests.post(url=token_url,
+                    data={
+                        'grant_type': 'authorization_code',
+                        'client_id': self.client_id,
+                        'redirect_uri': redirect_uri,
+                        'code': codestring,
+                        'code_verifier': code_verifier
+                    }
+                )
+            result = resp.json()
+            token_dict = {
+                'access_token': result['access_token'],
+                'token_type': 'bearer'
+            }
+            oauth = OAuth2Session(client_id=self.client_id, token=token_dict)
+            # oauth.fetch_token(token_url, authorization_response=authorization_url,code=codestring)
+            logger.debug("Token received: ")
+            logger.debug(oauth)
+            logger.debug("Start serial")
+            if token_file != None:
+                self._serializeToken(oauth.token,token_file)
+                logger.info("Token serialized to "+token_file)
+            logger.info("New token created")
+            #TODO throw an exception if oauth is null and implement the auth required method
+            # RELLY TODO AttributeError
+            return oauth
+
+        # TODO tranform to exception
+        
+    def renewToken(self):
+        logger.info("Token expired, renewing")
+        self.oauth=self.__getNewToken(self.username,self.password,self.token_file)
+        logger.info("Token renewed successfully")
+            
+        
+    """Get URL using OAuth session. Automatically renew the token if needed
```

Comment:
A quite random note, feel free to ignore, but it's pretty weird for python to see a docstring above a function name.